f3.go

```diff
@@ -145,412 +78,197 @@ func New(ctx context.Context, manifest manifest.ManifestProvider, ds datastore.D
-		return xerrors.Errorf("getting initial power tipset: %w", err)
+	if cs == nil {
+		return nil, xerrors.Errorf("F3 is not running")
 	}
+	return cs.Get(ctx, instance)
+}
 
-	initialPowerTable, err := m.ec.GetPowerTable(ctx, ts.Key())
-	if err != nil {
-		return xerrors.Errorf("getting initial power table: %w", err)
+// Returns the time at which the F3 instance specified by the passed manifest should be started, or
+// 0 if the passed manifest is nil.
+func (m *F3) computeBootstrapDelay(manifest *manifest.Manifest) (delay time.Duration, err error) {
+	if manifest == nil {
+		return 0, nil
 	}
 
-	cs, err := certstore.CreateStore(ctx, m.ds, m.Manifest().InitialInstance, initialPowerTable)
+	ts, err := m.ec.GetHead(m.runningCtx)
 	if err != nil {
-		return xerrors.Errorf("creating certstore: %w", err)
-	}
-	m.setCertStore(cs)
-	return nil
-}
-
-func (m *F3) GetLatestCert(ctx context.Context) (*certs.FinalityCertificate, error) {
-	m.mLk.Lock()
-	defer m.mLk.Unlock()
-	if m.certStore == nil {
-		return nil, xerrors.Errorf("F3 is not running")
-	}
-	return m.certStore.Latest(), nil
-}
-func (m *F3) GetCert(ctx context.Context, instance uint64) (*certs.FinalityCertificate, error) {
-	m.mLk.Lock()
-	defer m.mLk.Unlock()
-	if m.certStore == nil {
-		return nil, xerrors.Errorf("F3 is not running")
-	}
-	return m.certStore.Get(ctx, instance)
+		err := xerrors.Errorf("failed to get the EC chain head: %w", err)
+		return 0, err
+	}
+
+	// We bootstrap now if we're at the correct time, whether or
+	// not we've hit the correct epoch. The important thing is
+	// that:
+	//
+	// 1. All nodes will use BootstrapEpoch - Finality to pick the base.
+	// 2. All nodes will bootstrap at the same time.
+	currentEpoch := ts.Epoch()
+	if currentEpoch < manifest.BootstrapEpoch {
+		epochDelay := manifest.BootstrapEpoch - currentEpoch
+		start := ts.Timestamp().Add(time.Duration(epochDelay) * manifest.ECPeriod)
+		return max(time.Until(start), 0), nil
+	}
+	return 0, nil
 }
 
 // Run start the module. It will exit when context is cancelled.
 // Or if there is an error from the message handling routines.
-func (m *F3) Run(ctx context.Context) (_err error) {
-	ctx, m.cancelCtx = context.WithCancel(ctx)
-	defer m.cancelCtx()
-
-	runnerErrCh := make(chan error, 1)
-	manifestErrCh := make(chan error, 1)
-
-	// bootstrap runner for the initial manifest
-	go m.startGpbftRunner(ctx, runnerErrCh)
-
-	// run manifest provider. This runs a background goroutine that will
-	// handle dynamic manifest updates if this is a dynamic manifest provider.
-	// If it is a static manifest it does nothing.
-	go m.client.manifest.Run(ctx, manifestErrCh)
-
-	// teardown pubsub on shutdown
-	defer func() {
-		teardownErr := m.teardownPubsub(m.Manifest())
-		_err = multierr.Append(_err, teardownErr)
-	}()
-
-	select {
-	case <-ctx.Done():
-		return nil
-	case err := <-runnerErrCh:
-		return err
-	case err := <-manifestErrCh:
+func (m *F3) Start(startCtx context.Context) (_err error) {
```

Comment:
So:

- I like the start pattern because I can easily wait and return errors for setup issues.
- I don't like the start pattern because I get no notifications that something went wrong.

I'm thinking we should add a `Wait` to all of these services? Or something like that?

Reply:
Basically:

```go
func (m *F3) WaitStop(ctx context.Context) error {
    select {
    case <-m.runningCtx.Done(): // should exit when everything is done.
    case <-ctx: // hm... return immediately? but that's a bit confusing.
    }
    return m.Close()
}
```

Reply:
yeha, I'm not sure, Run flow also has its issues

Reply:
Even better, `func Done() <-chan struct{}`.